common/flatpak-utils.c

```diff
@@ -6132,3 +6132,112 @@ flatpak_show_cursor (void)
 {
   write (STDOUT_FILENO, FLATPAK_ANSI_SHOW_CURSOR, strlen (FLATPAK_ANSI_SHOW_CURSOR));
 }
+
+/* Expand variable references in @unexpanded, replacing with variables
+ * from @vars or the environment. We understand $VAR, ${VAR}, ${VAR:-WORD}
+ * and ${VAR:+WORD} as well as \$, \\
+ * Variable names must start with an alphabetic char and contain alphanumerics
+ * and '_'
+ */
+char *
+flatpak_expand_env_vars (const char  *unexpanded,
+                         char       **env)
+{
+  g_autoptr(GString) s = g_string_sized_new (256);
+  const char *p = unexpanded;
+
+  while (*p)
+    {
+      if (p[0] == '\\')
+        {
+          p++;
+          if (p[0] != '\\' && p[0] != '$')
+            g_string_append_c (s, '\\');
+          g_string_append_c (s, p[0]);
+          p++;
+          continue;
+        }
+
+      if (p[0] == '$' && (g_ascii_isalpha (p[1]) || (p[1] == '{' && g_ascii_isalpha (p[2]))))
+        {
+          g_autofree char *parameter = NULL;
+          g_autofree char *word = NULL;
+          const char *p0;
+          const char *q;
+          const char *value;
+          const char *r;
+
+          p++;
+          p0 = q = p;
+          if (*p0 == '{')
+            p0++;
+          p = p0;
+          while (*p == '_' || g_ascii_isalnum (*p))
+            p++;
+          parameter = g_strndup (p0, p - p0);
+          if (*q == '{')
+            {
+              if (p[0] == ':' && (p[1] == '-' || p[1] == '+'))
+                {
+                 int bal = 1;
+
+                  p++;
+                  r = p;
+                  p++;
+                  while (*p != '\0')
+                    {
+                      /* FIXME: escaping should be possible here */
+                      if (p[0] == '$' && p[1] == '{') bal++;
+                      if (p[0] == '}') bal--;
+                      if (bal == 0) break;
+                      p++;
+                    }
+                  word = g_strndup (r, p - r);
+                }
+              else if (p[0] != '}')
+                {
+                  g_autofree char *ref = g_strndup (p0, p - p0);
+                  g_warning ("Can't parse variable reference ${%s", ref);
+                  g_string_append (s, "${");
+                  g_string_append (s, ref);
+                  continue;
+                }
+            }
+
+          value = g_environ_getenv (env, parameter);
+          if (value == NULL)
+            value = g_getenv (parameter);
```

Comment:
This is wrong. We should not expand inside the sandbox an env-var that will not get set in the environment. What we want is a simple way to avoid a wrapper script in the app, not a way to leak host env vars into the sandbox.

Reply:
Our docs say:

> Environment variables are generally passed on to the sandboxed application, with certain exceptions.

Reply:
For generic runs, not not for e.g. builds, and some variables are special. This happens when we construct the  FlatpakBwrap though, where we initialize bwrap->envp with the real environment in these cases. After that we should never touch the real environment.

Reply:
This is still wrong. At this point we should have imported every host env var value we're gonna export to the sandbox into envp, so we should no more look at the external environment here.